coreax/coresubset.py

```diff
@@ -407,3 +407,177 @@ def fit_to_size(self, coreset_size: int) -> None:
+
+        except IndexError as exception:
+            if coreset_size == 0:
+                raise ValueError("coreset_size must be non-zero") from exception
+            raise
+
+        self.coreset_indices = coreset_indices
+        self.coreset = self.original_data.pre_coreset_array[self.coreset_indices, :]
+
+    @staticmethod
+    @partial(jit, static_argnames=["kernel_vectorised", "unique"])
+    def _loop_body(
+        i: int,
+        val: tuple[ArrayLike, ArrayLike, ArrayLike, coreax.util.KeyArrayLike],
+        x: ArrayLike,
+        kernel_vectorised: coreax.util.KernelComputeType,
+        unique: bool,
+    ) -> tuple[Array, Array, Array, Array]:
+        r"""
+        Execute main loop of RPCHOLESKY.
+
+        This function carries out one iteration of RPCHOLESKY, defined in Algorithm 1 of
+        :cite:`chen2023randomly`.
+
+        :param i: Loop counter, counting how many points are in the coreset on call of
+            this method
+        :param val: Tuple containing a :math:`m \times 1` array of the residual
+            diagonal, :math:`n \times m` Cholesky matrix F, :math:`1 \times m` array of
+            current coreset indices and a PRNGKey for sampling. Note that the array
+            holding current coreset indices should always be the same length (however
+            many coreset points are desired). The ``i``-th element of this gets updated
+            to the index of the selected coreset point in iteration ``i``.
+        :param x: :math:`n \times d` data matrix
+        :param kernel_vectorised: Vectorised kernel computation function. This should be
+            the :meth:`~coreax.kernel.Kernel.compute` method of a
+            :class:`~coreax.kernel.Kernel` object
+        :param unique: Flag for enforcing unique elements
+        :returns: Updated loop variables ``residual_diagonal``, ``F``,
+            ``current_coreset_indices`` and ``key``
+        """
+        # Unpack the components of the loop variables
+        residual_diagonal, approximation_matrix, current_coreset_indices, key = val
+        key, subkey = random.split(key)
+        n = len(x)
+
+        # sample a new index with probability proportional to residual diagonal
+        j = random.choice(subkey, n, (1,), p=residual_diagonal, replace=False)[0]
+        current_coreset_indices = current_coreset_indices.at[i].set(j)
+
+        # remove overlap with previously chosen columns
+        g = (
+            kernel_vectorised(x, x[j])
+            - jnp.dot(approximation_matrix, approximation_matrix[j])[:, None]
+        )
+
+        # update approximation
+        approximation_matrix = approximation_matrix.at[:, i].set(
+            (g / jnp.sqrt(g[j])).flatten()
+        )
+
+        # track diagonal of residual matrix
+        residual_diagonal = residual_diagonal - jnp.square(approximation_matrix[:, i])
+
+        # ensure diagonal remains nonnegative
+        residual_diagonal = residual_diagonal.clip(min=0)
+        if unique:
```

Comment:
All unit tests for RPCholesky have this condition as true - I suggest we add one test where it's false to cover that case (since we track coverage, and want to keep the entire codebase hit with tests)

Reply:
Done.